src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,557 @@
+ * type are provided: {@link Style#FULL FULL}, {@link Style#SHORT SHORT}, and
+ * {@link Style#NARROW NARROW}. The following snippet is an example of formatting
+ * the list of Strings {@code ["Foo", "Bar", "Baz"]} in US English with
+ * {@code STANDARD}, {@code FULL} (typical case):
+ * {@snippet lang=java :
+ * ListFormat.getInstance(Locale.US, ListFormat.Type.STANDARD, ListFormat.Style.FULL)
+ *     .format(List.of("Foo", "Bar", "Baz"))
+ * }
+ * This will produce the concatenated list string, "Foo, Bar, and Baz" as seen in
+ * the following table:
+ * <table class="striped">
+ * <caption style="display:none">Formatting examples</caption>
+ * <thead>
+ * <tr><th scope="col"></th>
+ *     <th scope="col">FULL</th>
+ *     <th scope="col">SHORT</th>
+ *     <th scope="col">NARROW</th></tr>
+ * </thead>
+ * <tbody>
+ * <tr><th scope="row" style="text-align:left">STANDARD</th>
+ *     <td>Foo, Bar, and Baz</td>
+ *     <td>Foo, Bar, &amp; Baz</td>
+ *     <td>Foo, Bar, Baz</td>
+ * <tr><th scope="row" style="text-align:left">OR</th>
+ *     <td>Foo, Bar, or Baz</td>
+ *     <td>Foo, Bar, or Baz</td>
+ *     <td>Foo, Bar, or Baz</td>
+ * <tr><th scope="row" style="text-align:left">UNIT</th>
+ *     <td>Foo, Bar, Baz</td>
+ *     <td>Foo, Bar, Baz</td>
+ *     <td>Foo Bar Baz</td>
+ * </tbody>
+ * </table>
+ * <p>
+ * Alternatively, Locale, Type, and/or Style invariant patterns can be
+ * specified with {@link #getInstance(String[])}. The String array to the
+ * method specifies the delimiting patterns for the start/middle/end portion of
+ * the formatted string, as well as optional specialized patterns for two or three
+ * elements. Refer to the method description for more detail.
+ *
+ * @implSpec This class is immutable and thread-safe
+ *
+ * @since 22
+ */
+public class ListFormat extends Format {
+
+    @Serial
+    private static final long serialVersionUID = 5272525550078071946L;
+
+    private static final int START = 0;
+    private static final int MIDDLE = 1;
+    private static final int END = 2;
+    private static final int TWO = 3;
+    private static final int THREE = 4;
+    private static final int PATTERN_ARRAY_LENGTH = THREE + 1;
+
+    /**
+     * The locale to use for formatting list patterns.
+     * @serial
+     */
+    private final Locale locale;
+
+    /**
+     * The array of five pattern Strings. Each element corresponds to the Unicode LDML's
+     * `listPatternsPart` type, i.e, start/middle/end/two/three.
+     * @serial
+     */
+    private final String[] patterns;
+
+    private transient Pattern startPattern;
+    private transient String middleBetween;
+    private transient Pattern endPattern;
+
+    private ListFormat(Locale l, String[] patterns) {
+        locale = l;
+        this.patterns = patterns;
+        init();
+    }
+
+    private void init() {
+        // get pattern strings
+        var m = Pattern.compile("(?<startBefore>.*?)\\{0}(?<startBetween>.*?)\\{1}").matcher(patterns[START]);
```

Comment:
Each of the patterns should be static, they don't need to be re-compiled for each ListFormat created.